clap_derive/src/derives/subcommand.rs

```diff
@@ -49,62 +49,72 @@ fn gen_augment_subcommands(
-                    Unit => quote!( #app_var ),
+
+                Kind::Flatten => match variant.fields {
                     Unnamed(FieldsUnnamed { ref unnamed, .. }) if unnamed.len() == 1 => {
                         let ty = &unnamed[0];
-                        quote_spanned! { ty.span()=>
-                            {
-                                <#ty as ::clap::IntoApp>::augment_clap(#app_var)
-                            }
+                        quote! {
+                            let app = <#ty as ::clap::Subcommand>::augment_subcommands(app);
                         }
                     }
-                    Unnamed(..) => {
-                        abort!(variant, "non single-typed tuple enums are not supported")
-                    }
-                };
+                    _ => abort!(
+                        variant,
+                        "`flatten` is usable only with single-typed tuple variants"
+                    ),
+                },
 
-                let name = attrs.cased_name();
-                let from_attrs = attrs.top_level_methods();
-                let version = attrs.version();
-                quote! {
-                    let app = app.subcommand({
-                        let #app_var = ::clap::App::new(#name);
-                        let #app_var = #arg_block;
-                        #app_var#from_attrs#version
-                    });
+                _ => {
+                    let app_var = Ident::new("subcommand", Span::call_site());
+                    let arg_block = match variant.fields {
+                        Named(ref fields) => {
+                            into_app::gen_app_augmentation(&fields.named, &app_var, &attrs)
+                        }
+                        Unit => quote!( #app_var ),
+                        Unnamed(FieldsUnnamed { ref unnamed, .. }) if unnamed.len() == 1 => {
+                            let ty = &unnamed[0];
+                            quote_spanned! { ty.span()=>
+                                {
+                                    <#ty as ::clap::IntoApp>::augment_clap(#app_var)
+                                }
+                            }
+                        }
+                        Unnamed(..) => {
+                            abort!(variant, "non single-typed tuple enums are not supported")
+                        }
+                    };
+
+                    let name = attrs.cased_name();
+                    let from_attrs = attrs.top_level_methods();
+                    let version = attrs.version();
+                    quote! {
+                        let app = app.subcommand({
+                            let #app_var = ::clap::App::new(#name);
+                            let #app_var = #arg_block;
+                            #app_var#from_attrs#version
+                        });
+                    }
                 }
             }
-        }
-    });
+        })
+        .collect();
```

Comment:
I don't think you need the collect here.

Reply:
I need collect here so this mapping is executed right now, not lazily. I need it because it sets `ext_subcmd` and the code below relies on it.